### PR TITLE
Phoenix-5652 Improve IndexScrutinyTool perf by using SkipScan in index table queries

### PR DESCRIPTION
**Pull request by gokceni**

### CODE REVIEW

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        list.add(new Object[] {
+                "CREATE TABLE %s (ID INTEGER NOT NULL, NAME VARCHAR NOT NULL, ZIP INTEGER, EMPLOY_DATE TIMESTAMP, EMPLOYER VARCHAR CONSTRAINT PK_1 PRIMARY KEY (ID, NAME)) SALT_BUCKETS=2",
+                null, "CREATE INDEX %s ON %s (EMPLOY_DATE, ZIP) INCLUDE (EMPLOYER) " });
+        list.add(new Object[] {
+                "CREATE TABLE %s (COL1 VARCHAR(15) NOT NULL,ID INTEGER NOT NULL, NAME VARCHAR, ZIP INTEGER CONSTRAINT PK_1 PRIMARY KEY (COL1, ID)) MULTI_TENANT=true",
+                null, "CREATE INDEX %s ON %s (NAME) " });
+        list.add(new Object[] {
+                "CREATE TABLE %s (ID INTEGER NOT NULL PRIMARY KEY, NAME VARCHAR, ZIP INTEGER, EMPLOY_DATE TIMESTAMP, EMPLOYER VARCHAR)",
+                "CREATE VIEW %s AS SELECT * FROM %s", "CREATE INDEX %s ON %s (NAME) " });
+//        list.add(new Object[] {
```
Comment (swaroopak) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: nit: please remove the commented code.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/GlobalIndexCheckerIT.java [hunk 2]
```diff
@@ -300,7 +300,6 @@ public void testOnePhaseOverwiteFollowingTwoPhaseWrite() throws Exception {
             verifyTableHealth(conn, dataTableName, indexTableName);
         }
     }
-
```
Comment (swaroopak) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/index/GlobalIndexCheckerIT.java]: nit: unnecessary diff

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/IndexColumnNames.java [hunk 3]
```diff
@@ -252,4 +288,61 @@ public String getQualifiedIndexTableName() {
...
+   //     return indexPkColNamesForSkipScan;
+    }
+
+    public List<String> getIndexColNamesForSkipScan() {
+        if (pdataTable.isMultiTenant()) {
+            return indexColNames;
+        } else {
+            return indexColNamesForSkipScan;
+        }
+ //      return indexColNamesForSkipScan;
```
Comment (swaroopak) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/IndexColumnNames.java]: nit: please remove the commented code.

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/IndexColumnNames.java [hunk 4]
```diff
@@ -252,4 +288,61 @@ public String getQualifiedIndexTableName() {
...
+ //       return dataColNamesForSkipScan;
+    }
+
+    public List<String> getIndexPkColNamesForSkipScan() {
+        if (pdataTable.isMultiTenant()) {
+            return indexPkColNames;
+        } else {
+            return indexPkColNamesForSkipScan;
+        }
+   //     return indexPkColNamesForSkipScan;
```
Comment (swaroopak) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/IndexColumnNames.java]: nit: please remove the commented code.

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/IndexColumnNames.java [hunk 5]
```diff
@@ -252,4 +288,61 @@ public String getQualifiedIndexTableName() {
...
         return indexPkColNames;
     }
+
+    public List<String> getDataColNamesForSkipScan() {
+        if (pdataTable.isMultiTenant()) {
+            return dataColNames;
+        } else {
+            return dataColNamesForSkipScan;
+        }
+ //       return dataColNamesForSkipScan;
```
Comment (swaroopak) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/IndexColumnNames.java]: nit: please remove the commented code.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    public IndexScrutinyToolPerformanceIT(String dataTableDdl, String viewDdl, String indexTableDdl) {
+        this.dataTableDdl = dataTableDdl;
+        this.viewDdl = viewDdl;
+        this.indexTableDdl = indexTableDdl;
+    }
+
+    /**
+     * Create the test data and index tables
+     */
+    @Before public void setup() throws SQLException {
```
Comment (swaroopak) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: nit: start the access modifiers on the new line

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolIT.java [hunk 7]
```diff
@@ -694,7 +702,6 @@ private void upsertRow(PreparedStatement stmt, int id, String name, byte[] val)
         stmt.setInt(index++, id);
         stmt.setString(index++, name);
         stmt.setBytes(index++, val);
-        stmt.executeUpdate();
```
Comment (swaroopak) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolIT.java]: why was this removed?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+    private Connection conn;
+
+    private PreparedStatement dataTableUpsertStmt;
+    private PreparedStatement viewUpsertStmt;
+
+    private long testTime;
+    private Properties props;
+
+    private static final Log LOG = LogFactory.getLog(IndexScrutinyToolPerformanceIT.class);
```
Comment (swaroopak) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: nit: Let's be consistent with using LOGGER and also the corresponding import

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+        if (!Strings.isNullOrEmpty(viewDdl)) {
+            baseTableName = viewName;
+            upsertStmt = viewUpsertStmt;
+        }
+
+        for (int i=0; i < 2; i++) {
+            if (dataTableDdl.contains("MULTI_TENANT")) {
+                upsertRowMultiTenant(upsertStmt, "val"+i, i,"name-" + i, 19999);
+            } else {
+                upsertRow(upsertStmt, i, "name-" + i, 19999);
```
Comment (swaroopak) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: if I understand this correctly, this ll only insert 2 rows?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+        PreparedStatement targetStmt = IndexScrutinyTool.getTargetTableQueryForScrutinyTool();
+        assertSkipScanFilter(targetStmt);
+
+        IndexScrutinyToolIT.runScrutiny(schemaName, baseTableName,
+                indexTableName, null, SourceTable.DATA_TABLE_SOURCE);
+        targetStmt = IndexScrutinyTool.getTargetTableQueryForScrutinyTool();
+        assertSkipScanFilter(targetStmt);
+    }
+
+    private void assertSkipScanFilter(PreparedStatement targetStmt) throws SQLException {
```
Comment (swaroopak) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: Shall we have a test for measuring performance as well?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+        int index = 1;
+        // insert row
+        stmt.setInt(index++, id);
+        stmt.setString(index++, name);
+        stmt.setInt(index++, zip);
+        stmt.setTimestamp(index++, new Timestamp(testTime));
+        stmt.executeUpdate();
+    }
+
+    private void upsertRowMultiTenant(PreparedStatement stmt, String col, int id, String name, int zip) throws SQLException {
```
Comment (swaroopak) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: Can this be moved to IndexScrutinyToolBaseIT and reused in both ITs, this and IndexScrutinyToolIT?

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/IndexColumnNames.java [hunk 12]
```diff
@@ -252,4 +288,61 @@ public String getQualifiedIndexTableName() {
...
+
+        for (String colName : indexColNamesForSkipScan) {
+            indexSqlTypes.add(indexColSqlTypeNamesForSkipScan.get(colName));
+        }
+
+        return getDynamicCols(getUnqualifiedColNames(indexColNamesForSkipScan), indexSqlTypes);
+    }
+
+
+    public List<String> getDynamicDataColsForSkipScan() {
```
Comment (swaroopak) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/IndexColumnNames.java]: good to have a test on dynamic columns as well.

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/IndexColumnNames.java [hunk 13]
```diff
@@ -79,6 +85,8 @@ public IndexColumnNames(final PTable pdataTable, final PTable pindexTable) {
                 dataColSqlTypeNames.add(getDataTypeString(dPkCol));
                 indexPkColNames.add(indexColumnName);
                 indexColSqlTypeNames.add(getDataTypeString(indexCol));
+                indexColSqlTypeNamesForSkipScan.put(indexColumnName, getDataTypeString(dPkCol));
```
Comment (swaroopak) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/IndexColumnNames.java]: Is using skipscan for indexscrutiny optional? If not, can the above two lists be merged into?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java (hunk 9 above)
Comment (gokceni) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: yes, we just check if the filter is used or not.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java (hunk 10 above)
Comment (gokceni) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: Not sure how we can do that

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/IndexColumnNames.java (hunk 13 above)
Comment (gokceni) [reply phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/IndexColumnNames.java]: It is not optional but I need these both.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java (hunk 11 above)
Comment (gokceni) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: IndexScrutinyToolIT doesn't have multi tenant test. There is a separate test file for multi tenant but it is a bigger change. It doesn't use prepared statement and uses tenant connections and global connections.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolIT.java [hunk 14]
```diff
@@ -165,19 +167,23 @@ public IndexScrutinyToolIT(String dataTableDdl, String indexTableDdl) {
     @Test public void testScrutinyOnArrayTypes() throws Exception {
         String dataTableName = generateUniqueName();
         String indexTableName = generateUniqueName();
-        String dataTableDDL = "CREATE TABLE %s (ID INTEGER NOT NULL PRIMARY KEY, NAME VARCHAR, VB VARBINARY)";
+        String
+                dataTableDDL =
```
Comment (gjacoby126) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolIT.java]: nit: any reason dataTableDDL is on its own line?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolIT.java [hunk 15]
```diff
@@ -165,19 +167,23 @@ public IndexScrutinyToolIT(String dataTableDdl, String indexTableDdl) {
...
             conn.createStatement().execute(String.format(dataTableDDL, dataTableName));
-            conn.createStatement().execute(String.format(indexTableDDL, indexTableName, dataTableName));
+            conn.createStatement()
+                    .execute(String.format(indexTableDDL, indexTableName, dataTableName));
             // insert two rows
-            PreparedStatement upsertDataStmt = conn.prepareStatement(String.format(upsertData, dataTableName));
-            upsertRow(upsertDataStmt, 1, "name-1", new byte[] {127, 0, 0, 1});
-            upsertRow(upsertDataStmt, 2, "name-2", new byte[] {127, 1, 0, 5});
+            PreparedStatement
+                    upsertDataStmt =
```
Comment (gjacoby126) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolIT.java]: ditto upsertDataStmt?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolIT.java [hunk 16]
```diff
@@ -188,13 +194,15 @@ public IndexScrutinyToolIT(String dataTableDdl, String indexTableDdl) {
             assertEquals(0, getCounterValue(counters, INVALID_ROW_COUNT));
 
             // Now insert a different varbinary row
-            upsertRow(upsertDataStmt, 3, "name-3", new byte[] {1, 1, 1, 1});
+            upsertRow(upsertDataStmt, 3, "name-3", new byte[] { 1, 1, 1, 1 });
```
Comment (gjacoby126) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolIT.java]: nit: unnecessary space

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolIT.java [hunk 17]
```diff
@@ -188,13 +194,15 @@ public IndexScrutinyToolIT(String dataTableDdl, String indexTableDdl) {
...
             conn.commit();
 
-            PreparedStatement upsertIndexStmt = conn.prepareStatement(String.format(upsertIndex, indexTableName));
+            PreparedStatement
+                    upsertIndexStmt =
```
Comment (gjacoby126) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolIT.java]: nit: ditto upsertIndexStmt

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolIT.java [hunk 18]
```diff
@@ -188,13 +194,15 @@ public IndexScrutinyToolIT(String dataTableDdl, String indexTableDdl) {
...
+                    conn.prepareStatement(String.format(upsertIndex, indexTableName));
             upsertIndexStmt.setString(1, "name-3");
             upsertIndexStmt.setInt(2, 3);
-            upsertIndexStmt.setBytes(3, new byte[] {0, 0, 0, 1});
+            upsertIndexStmt.setBytes(3, new byte[] { 0, 0, 0, 1 });
```
Comment (gjacoby126) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolIT.java]: nit: unnecessary space

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+                null, "CREATE LOCAL INDEX %s ON %s (NAME, EMPLOY_DATE) INCLUDE (ZIP)" });
+        list.add(new Object[] {
+                "CREATE TABLE %s (ID INTEGER NOT NULL PRIMARY KEY, NAME VARCHAR, ZIP INTEGER, EMPLOY_DATE TIMESTAMP, EMPLOYER VARCHAR)",
+                null, "CREATE INDEX %s ON %s (NAME, ID, ZIP, EMPLOYER, EMPLOY_DATE)" });
+        list.add(new Object[] {
+                "CREATE TABLE %s (ID INTEGER NOT NULL PRIMARY KEY, NAME VARCHAR, ZIP INTEGER, EMPLOY_DATE TIMESTAMP, EMPLOYER VARCHAR)",
+                null, "CREATE INDEX %s ON %s (NAME, ZIP, EMPLOY_DATE, EMPLOYER)" });
+        list.add(new Object[] {
+                "CREATE TABLE %s (ID INTEGER NOT NULL, NAME VARCHAR NOT NULL, ZIP INTEGER, EMPLOY_DATE TIMESTAMP, EMPLOYER VARCHAR CONSTRAINT PK_1 PRIMARY KEY (ID, NAME)) ",
+                null, "CREATE INDEX %s ON %s (EMPLOY_DATE) INCLUDE (ZIP)" });
```
Comment (gjacoby126) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: nit: normally I don't flag long lines warnings very often -- but these lines are really long.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java [hunk 20]
```diff
@@ -0,0 +1,246 @@
...
+    private PreparedStatement viewUpsertStmt;
+
+    private long testTime;
+    private Properties props;
+
+    private static final Log LOGGER = LogFactory.getLog(IndexScrutinyToolPerformanceIT.class);
+    @Parameterized.Parameters public static Collection<Object[]> data() {
+        List<Object[]> list = Lists.newArrayListWithExpectedSize(15);
+        list.add(new Object[] {
+                "CREATE TABLE %s (ID INTEGER NOT NULL PRIMARY KEY, NAME VARCHAR, ZIP INTEGER, EMPLOY_DATE TIMESTAMP, EMPLOYER VARCHAR)",
```
Comment (gjacoby126) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: Some comments about what each parameter is testing would be useful, because it's not always immediately obvious.

Comment (gjacoby126) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: Also, given that we're adding so many cases here, are there redundancies between this and the other IndexScrutinyIT and related tests that can be consolidated? No need to test the same scenarios twice.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+                upsertRowMultiTenant(upsertStmt, "val"+i, i,"name-" + i, 19999);
+            } else {
+                upsertRow(upsertStmt, i, "name-" + i, 19999);
+            }
+            conn.commit();
+        }
+
+        List<Job> completedJobs = IndexScrutinyToolIT.runScrutiny(schemaName, baseTableName,
+                indexTableName, null, SourceTable.DATA_TABLE_SOURCE);
+        PreparedStatement targetStmt = IndexScrutinyTool.getTargetTableQueryForScrutinyTool();
```
Comment (gjacoby126) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: What's the goal of this test? If it's just to validate that we're using skip scan filters when appropriate, is there a point to running the scrutiny? (Or at least, to running it for every test case?) If there is, as I asked above, are there scrutiny test runs elsewhere that are no longer necessary?

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapper.java [hunk 22]
```diff
@@ -54,6 +58,8 @@
 
 import com.google.common.base.Joiner;
 
+import static org.junit.Assert.assertEquals;
```
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapper.java]: JUnit classes generally shouldn't be imported into non-test code, plus this looks unused.

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/IndexColumnNames.java [hunk 23]
```diff
@@ -61,8 +67,13 @@ public IndexColumnNames(final PTable pdataTable, final PTable pindexTable) {
         if (pindexTable.getViewIndexId() != null) {
             offset++;
         }
-        if (pindexTable.isMultiTenant() && pindexTable.getViewIndexId() != null) {
-            offset++;
+
+        if (tenantId != null) {
+            // For multi tenant view indexes, in order for skipscan to be used, we need tenant column.
+            // For tenant connections this column is not visible.
+            if (pindexTable.isMultiTenant() && pindexTable.getViewIndexId() != null) {
```
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/IndexColumnNames.java]: @gokceni  Wouldn't you also need this for multi-tenant global indexes not on a view?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java (hunk 21 above)
Comment (gjacoby126) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: If we're just running scrutiny to generate the statement we're looking up, should there be a dry run mode that just generates that statement but doesn't actually run the job?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java [hunk 24]
```diff
@@ -0,0 +1,246 @@
...
+        job = completedJobs.get(0);
+        counters = job.getCounters();
+        assertEquals(2, getCounterValue(counters, VALID_ROW_COUNT));
+        assertEquals(0, getCounterValue(counters, INVALID_ROW_COUNT));
+    }
+
+    private void assertSkipScanFilter(PreparedStatement targetStmt) throws SQLException {
+        QueryPlan plan = targetStmt.unwrap(PhoenixStatement.class).getQueryPlan();
+        Scan scan = plan.getScans().get(0).get(0);
+        Filter filter = scan.getFilter();
```
Comment (gjacoby126) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolPerformanceIT.java]: Is the presence of a SkipScan filter sufficient to guarantee correctness, or is there some other additional information we need, like what's being filtered?

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyTool.java [hunk 25]
```diff
@@ -259,18 +266,20 @@ public Job createSubmittableJob(String schemaName, String indexTable, String dat
...
+                                    pindexTable, tenantId)
                             : new SourceTargetColumnNames.IndexSourceColNames(pdataTable,
-                                    pindexTable);
+                                    pindexTable, tenantId);
             String qSourceTable = columnNames.getQualifiedSourceTableName();
-            List<String> sourceColumnNames = columnNames.getSourceColNames();
-            List<String> sourceDynamicCols = columnNames.getSourceDynamicCols();
-            List<String> targetDynamicCols = columnNames.getTargetDynamicCols();
+
+            List<String> sourceColumnNames = columnNames.getSourceColNamesForSkipScan();
```
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyTool.java]: Seems like we're "leaking" implementation details here...is there a reason why the job creator function needs to know the Scan's going to be a skip scan? Why should it need to change in the future if we decide to use a different Scan filter?

**Comment by gokceni** [comment]
Closing this PR because we did not see that much improvement in performance as we were hoping to see. It is still faster than current version but not fast enough. The -b 1 is much faster than this option
